src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -0,0 +1,51 @@
+require 'puppet_pal'
+require 'puppet/server/log_collector'
+require 'puppet/server/logging'
+
+module Puppet
+  module Server
+    class ASTCompiler
+      def self.compile(compile_options)
+        options = compile_options['options'] || {}
+
+        log_level = options['log_level']
+
+        if options['capture_logs']
+          catalog, logs = Logging.capture_logs(log_level) do
+            compile_ast(compile_options)
+          end
+
+          { catalog: catalog, logs: logs }
+        else
+          catalog = compile_ast(compile_options)
+          { catalog: catalog }
+        end
+      end
+
+      def self.compile_ast(compile_options)
+        # FromDataConverter.convert below expects this to be a hash
+        code = JSON.parse(compile_options['code_ast'])
```

Comment:
It seems weird that fact values have been recursively deserialized but the ast hasn't at this point. Should the code higher up be doing some more aggressive deserializing?

Reply:
Yeah that seems fine. This is a relic from when I thought the processing methods below were expecting a string.